openedx/openedx-platform
# WIP: Add task to call clearsessions

This adds a celery task to call the `clearsessions` management command. This can be configured from celery's crontab to run periodically.
The goal is to avoid cases where there hundreds of thousands of expired sessions that slow down the DB. 

It allows you to use a parameter in `lms.env.json`, like `"CLEAR_SESSIONS_PERIOD_HOURS": 24`, to run it every 24 hours.
By default it's disabled.

*WIP* waiting for internal review first

**JIRA tickets**: None
**Discussions**:
- It is implemented as a celery task, not as a cron in `configuration`, to be able to do tests for it, and because most tasks in edX use this `tasks.py` format rather than cron.
- Please comment if you think there's a more sensible approach to clear sessions

**Dependencies**: None
**Screenshots**: None
**Sandbox URL**: None
**Merge deadline**: None

**Testing instructions**:

1. Run the tests. It can be in devstack or CI.
1. Provision a sandbox and use this branch (it's harder to test celery in a devstack)
1. Make sure that you use a `SESSION_ENGINE` variable that supports clearing sessions, like `django.contrib.sessions.backends.cached_db` (but not `….cache`)
```
>>> from django.conf import settings
>>> settings.SESSION_ENGINE
u'django.contrib.sessions.backends.cached_db'
```
1. In your `lms.env.json`, add `"CLEAR_SESSIONS_PERIOD_HOURS": 0.033,` (this is the period at which to run the command; that value is 2 minutes)
1. Restart `edxapp:` and `edxapp_worker:` through supervisor
1. Watch the log: `tail -f /edx/var/log/edx.log`
1. As edxapp user, start the celery beat: `/edx/app/edxapp/venvs/edxapp/bin/python /edx/app/edxapp/edx-platform/manage.py lms --settings=aws celery beat`
1. Watch the log (and the `beat` log). Every 2 minutes you should see something like:
  ```
Jan 29 13:44:17 edxapp-pr16640sandbo-appserver-1 [service_variant=lms][celery.worker.job][env:sandbox] INFO [edxapp-pr16640sandbo-appserver-1  9542] [job.py:401] - Task util.tasks.run_clearsessions[d52f3ff3-4ea3-4e88-89ad-7a781fb4d1c9] succeeded in 0.0154942395166s: None
```
1. Create a very old session and watch it be deleted. For this, run this script in `./manage.py lms shell --settings=aws`:
```
from django.contrib.sessions.backends.db import SessionStore as DatabaseSession
from django.contrib.sessions.models import Session
Session.objects.count()
other_session = DatabaseSession()
other_session.set_expiry(-3600)
other_session.save()
Session.objects.count()
# wait around 2 minutes for the Celery task to run
Session.objects.count()
```
1. You should see it go from 0 to 1 and then back to 0. Actually it should do the same as `./manage.py lms clearsessions`. Note that if you have your browser open and refreshing, you might be unwillingly creating a new session. Also note that other background activity creates new sessions (e.g. some code from `openedx.core.djangoapps.catalog.utils` was creating one every ~4 minutes).

**Reviewers**
- TBD edX

**Author concerns**: Still testing how to automatically start `celery beat`
**Settings**: None
